addons/web_editor/static/src/js/wysiwyg/PeerToPeer.js

```diff
@@ -0,0 +1,550 @@
+/** @odoo-module */
+
+window.showNotif = true;
+
+export class PeerToPeer {
+    constructor(options) {
+        this.options = options;
+        window.rtc = this;
+        this._currentClientId = this.options.currentClientId;
+
+        // clientId -> ClientInfos
+        this.clientsInfos = {};
+        this._lastRequestId = -1;
+        this._pendingRequestResolver = {};
+    }
+
+    getConnectedClientIds() {
+        return Object.entries(this.clientsInfos)
+            .filter(
+                ([id, infos]) =>
+                    infos.peerConnection.iceConnectionState === 'connected' &&
+                    infos.dataChannel.readyState === 'open',
+            )
+            .map(([id]) => id);
+    }
+
+    removeClient(clientId, { shouldNotify = true } = {}) {
+        console.log(`%c REMOVE CLIENT ${clientId}`, 'background: chocolate;');
+        this.notifySelf('ptp_remove', clientId);
+        const clientInfos = this.clientsInfos[clientId];
+        if (!clientInfos) return;
+        clearTimeout(clientInfos.fallbackTimeout);
+        clearTimeout(clientInfos.zombieTimeout);
+        clientInfos.dataChannel.close();
+        clientInfos.peerConnection.close();
+        delete this.clientsInfos[clientId];
+    }
+
+    async closeAllConnections() {
+        for (const clientId of Object.keys(this.clientsInfos)) {
+            await this.notifyAllClients('ptp_disconnect');
+            this.removeClient(clientId);
+        }
+    }
+
+    async notifyAllClients(notificationName, notificationPayload, { transport = 'server' } = {}) {
+        const transportPayload = {
+            fromClientId: this._currentClientId,
+            notificationName,
+            notificationPayload,
+        };
+        await this._simulateLatency(async () => {
+            if (transport === 'server') {
+                await this.options.broadcastAll(transportPayload);
+            } else if (transport === 'rtc') {
+                for (const cliendId of Object.keys(this.clientsInfos)) {
+                    // todo: Handle error if it happens.
```

Comment:
A quel point on doit faire ce todo avant de merger, idéalement ?

Reply:
At the end, it is not a necessity to handle error there. I've removed the todo.